File: Alg6_4.cpp
#include<iostream>
#include<vector>

using std::cout;
using std::cin;
using std::vector;

int N, M;


vector<int> visited;
vector<int> value;
vector<int> cycle;

int dfs(vector<vector<int>> graph, int now)
{
    if(visited[now] != 2)
    {
        visited[now] = 1;

            for(int i = 0; i < size(graph[now]); i++)
            {
                int neig = graph[now][i];

                if(visited[neig] == 0)
                {
                    if(!dfs(graph, neig)) return 0;
                    int count = 0;
                    for(int j = 0; j < size(value); j++)
                    {
                        if(value[j] == neig)
                        {
                            count++;
                        }
                    }
                    if(count == 0)
                    {
                        value.push_back(neig);
                    }
                    else count = 0;
                }
                else if(visited[neig] == 1) return 0;
                else continue;
                visited[neig] = 2;
            }

            int count = 0;
            for(int j = 0; j < size(value); j++)
            {
                if(value[j] == now)
                {
                    count++;
                }
            }
            if(count == 0)
            {
                value.push_back(now);
            }
            else count = 0;
            visited[now] = 2;
            return 1;
    }
    else if (visited[now] == 1) return 0;
    else return 1;
    
}

int main()
{

    cin >> N >> M;
    vector<vector<int>> graph(N + 1);
    for(int i = 1; i < M + 1; i++)
    {
        int value_1, value_2;
        cin >> value_1 >> value_2;
        graph[value_1].push_back(value_2);
    }
    visited.resize(N + 1);
    int count = 0;

    for(int i = 1; i < N + 1; i++)
    {
        if(!dfs(graph, i))
        {
            count++;
            break;
        }
    }

    if(count == 0)
    {
        for(int i = (size(value) - 1); i >= 0; i--)
        {
            cout << value[i] << " ";
        }
    }
    else cout << -1;
}

File: Alg6_4.py
from sys import setrecursionlimit

setrecursionlimit(40000)
value = []
def dfs(graph, visited, now = 1):
    if visited[now] != 2:
        visited[now] = 1
        for neig in graph[now]:
            if visited[neig] == 0:
                if dfs(graph, visited, neig) == 0:
                    return 0
                if neig not in value:
                    value.append(neig)
            elif visited[neig] == 1:
                return 0
            visited[neig] = 2

        if now not in value:
            value.append(now)
        visited[now] = 2
        return 1
    elif visited[now] == 1:
        return 0
    else:
        return 1

def main():

    values = 0
    with open("input.txt") as file:
        values = file.readlines()
        values = [i.rstrip() for i in values]


    NM = list(map(int, values[0].split(' ')))
    graph = [0] * (NM[0] + 1)
    for i in range(1, NM[0] + 1):
        graph[i] = []


    for i in range(1, NM[1] + 1):
        value = list(map(int, values[i].split(' ')))
        if value[1] not in graph[value[0]]:
            graph[value[0]].append(value[1])



    visited = [0] * (NM[0] + 1)

    for i in range(1, NM[0] + 1):
        if not dfs(graph, visited, i):
            return 0
    return 1



if __name__ == "__main__":
    if main():
        value = list(map(str, value))
        print(' '.join(value[::-1]))
    else:
        print(-1)